system/doc/design_principles/secure_coding.md

```diff
@@ -0,0 +1,1833 @@
+the same name-space, one can also unintentionally start an instance of a driver
+instead of spawning an external program.
+
+When spawning an external program, a safer approach is to use [`open_port/2`]
+with the `{spawn_executable, _}` argument. No shell is used and the arguments
+to the program needs to be explicitly listed using the `{args, Args}` option
+where no environment variable expansion is made. You may also want to overwrite
+certain environment variables using the `{env, Env}` option in order not to
+expose those to the external program.
+
+When starting an instance of a loaded driver, a safer approach is to use
+[`open_port/2`] with the `{spawn_driver, _}` argument where no mix-up with
+external programs may occur.
+
+External OS commands can also be executed by calling `os:cmd/1` (or
+`os:cmd/2`) with a command line as argument. `os:cmd/1,2` suffers from most of
+the same issues as [`open_port/2`] with the `{spawn, _}` argument. A shell will
+be started in which the passed command line is executed. The `PATH` will be
+searched for the command and environment variables will be expanded. A safer
+approach is to use [`open_port/2`] with the `{spawn_executable, _}` argument.
+It requires gathering of the output from the command instead of getting it
+returned as a string, but it is an easy task to handle.
+
+[`open_port/2`]: `erlang:open_port/2`
+
+[](){: #rule-msc-004 }
+[`MSC-004`]: #rule-msc-004
+#### [`MSC-004`] - Protect Sensitive Data
+
+As Erlang logs many things by default, it is common for application data to
+appear in log files and the like. This is not necessarily a problem but some
+environments are under regulatory or compliance requirements not to leak
+certain data to disk or similar in case an unauthorized person were to get
+ahold of them ([CWE-532]).
+
+As the system cannot make any distinction between which data is sensitive and
+which is not, it is up to the user to protect sensitive data. Ways of doing
+this include, but are not limited to:
+
+1. Using the `private` option for ETS tables containing sensitive data,
+   preventing the table contents from being read by other processes.
+1. Using `process_flag(sensitive, true)` for processes operating on sensitive
+   data.
+   
+   This disables nearly all introspection for the process: other processes
+   cannot inspect the message queue of this process, tracing is disabled, the
+   process' data will not be included in a crash dump, and so on.
+1. Implementing the `format_status/2` callback for `gen*` behaviors
+   (`m:gen_server` et al) for processes operating on sensitive data.
+   
+   This lets you control how the process state is presented in introspection
+   tools such as `observer`.
+
+Furthermore, to prevent leaking data by stack traces or certain introspection
+features, a common technique is to wrap secrets in a zero-arity `fun()` that is
+then called in order to retrieve the secret. When passed around, the `fun()`
```

Comment:
While this works, it prevents data from being passed around through dist erl (since the remote node might have a different version of the module) - this can be a severe limitation.

A way to still achieve this would be to wrap in a closure that always errors when called, but captures the value, and inspecting the value through `erlang:fun_info(F, env)` rather than calling it